asserts/pool.go

```diff
@@ -19,6 +19,538 @@
+		return err
+	}
+
+	if !a.SupportedFormat() {
+		return &UnsupportedFormatError{Ref: a.Ref(), Format: a.Format()}
+	}
+
+	uniq := a.Ref().Unique()
+	var u *unresolvedRec
+	var extrag *internal.Grouping
+	var unresolved map[string]*unresolvedRec
+	if u = p.unresolved[uniq]; u != nil {
+		unresolved = p.unresolved
+	} else if u = p.prerequisites[uniq]; u != nil {
+		unresolved = p.prerequisites
+	} else {
+		ok, err := p.isPredefined(a.Ref())
+		if err != nil {
+			return err
+		}
+		if ok {
+			// nothing to do
+			return nil
+		}
+		// a is not tracked as unresolved in any way so far,
+		// this is an atypical scenario where something gets
+		// pushed but we still want to add it to the resolved
+		// lists of the relevant groups; in case it is
+		// actually already resolved most of resolveWith below will
+		// be a nop
+		u = &unresolvedRec{
+			at: a.At(),
+		}
+		u.at.Revision = RevisionNotKnown
+	}
+
+	if u.label != grouping {
+		var err error
+		extrag, err = p.groupings.Parse(string(grouping))
+		if err != nil {
+			return err
+		}
+	}
+
+	return p.resolveWith(unresolved, uniq, u, a, extrag)
+}
+
+// TODO: AddBatch
+
+var ErrUnresolved = errors.New("unresolved assertion")
+
+// unresolvedBookkeeping processes any left over unresolved assertions
+// since the last ToResolve invocation and intervening calls to Add/AddBatch,
+//  * they were either marked as in error which will be propagated
+//    to all groups requiring them
+//  * simply unresolved, which will be propagated to groups requiring them
+//    as ErrUnresolved
+//  * unchanged (update case)
+// unresolvedBookkeeping will also promote any recorded prerequisites
+// into actively unresolved, as long as not all the groups requiring them
+// are in error.
+func (p *Pool) unresolvedBookkeeping() {
+	// any left over unresolved are either:
+	//  * in error
+	//  * unchanged
+	//  * or unresolved
+	for uniq, u := range p.unresolved {
+		e := u.err
+		if e == nil {
+			if u.at.Revision == RevisionNotKnown {
+				e = ErrUnresolved
+			}
+		}
+		if e != nil {
+			p.setErr(&u.grouping, e)
+		}
+		delete(p.unresolved, uniq)
+	}
+
+	// prerequisites will become the new unresolved but drop them
+	// if all their groups are in error
+	for uniq, prereq := range p.prerequisites {
+		useful := false
+		p.groupings.Iter(&prereq.grouping, func(gnum uint16) error {
+			if !p.groups[gnum].hasErr() {
+				useful = true
+			}
+			return nil
+		})
+		if !useful {
+			delete(p.prerequisites, uniq)
+			continue
+		}
+	}
+
+	// prerequisites become the new unresolved, the emptied
+	// unresolved is used for prerequisites in the next round
+	p.unresolved, p.prerequisites = p.prerequisites, p.unresolved
+}
+
+// Err returns the error for group if group is in error, nil otherwise.
+func (p *Pool) Err(group string) error {
+	gnum, err := p.groupNum(group)
+	if err != nil {
+		return err
+	}
+	gRec := p.groups[gnum]
+	if gRec == nil {
+		return fmt.Errorf("unknown group: %s", group)
```

Comment:
Nitpick, I think "unknown group" error could have a dedicated error type to distinguish it from actual group errors.

Reply:
introduced ErrUnknownPoolGroup

Reply:
Thanks!